### ISSUE TITLE
Numeric facet

### ISSUE THREAD

**Issue by fabiolinus** [OpenRefine/OpenRefine#5008]
Hi. Consider to open two text facets in openrefine. Second facet depends from the first facet (order is important). For example male and females showed in the second facet are those selected from the first choice (first facet). Suppose you choose a text facet, then a numeric facet. Numeric facet doesn't depend from the first facet. I think the numeric facet must change with the first facet. Thanks a lot for any help.

**Comment by thadguidry** [comment]
There is no dependency.  What you are witnessing is simply 2 facets that show values from filtered rows.  Each facet can have a different filtering effect.  However, there is a precedence order, in that if you move a facet into the top location of the facets panel, then that first facet will apply a filter to rows, then with those remaining filtered rows, the 2nd facet further filters, then the 3rd facet and so on.

If this is what you are seeing this is normal behavior.
If you can provide some sample data and steps to reproduce, than we can look into a potential bug around precedence ordering for numeric (or custom numeric) facets, but this is highly unlikely and instead you are simply seeing the above-described way facets work.

**Comment by fabiolinus** [comment]
Hi. Thanks for your effort, today is sunday! I made a GIF to clear my point of view. First I filter with a text facet, then I use a numeric facet. Data distributions changes with the first facet (see different levels of blue color). My dubt is that the scale remains the same (range 1-61). Why?

[data_numeric_facet.csv](https://github.com/OpenRefine/OpenRefine/files/9035173/data_numeric_facet.csv)
![NumericFacet](https://user-images.githubusercontent.com/33990409/177049482-82f754f0-f8e6-4a13-9fc5-fa78dbd44e69.gif)

**Comment by wetneb** [comment]
yes this is a valid point: the histogram is computed globally, on the entire dataset, whereas for a text facet the possible choices are computed only within the rows selected by the other facets. This is indeed a sort of inconsistency.
I did not design this, but I suspect that this was done to keep the bucketing of the numeric values independent from the facet options.

**Comment by fabiolinus** [comment]
Hi. I have a simple dataset:
A B1 a2 b3 c
I'd like to obtain a csv file with tow columns: A (numeric) and B (string):
A, "B"1, "a"2, "b"3, "c"
I use Custom Tabular Exporter but it puts quote evertwhere or ad a lot of quotes to the elements of column B:
""a""
or
"""a"""
How can I obtain a simple schema like before? Thanks a lot,
Fabio


    Il domenica 3 luglio 2022 14:41:55 CEST, Thad Guidry ***@***.***> ha scritto:  
 
 


There is no dependency. What you are witnessing is simply 2 facets that show values from filtered rows. Each facet can have a different filtering effect. However, there is a precedence order, in that if you move a facet into the top location of the facets panel, then that first facet will apply a filter to rows, then with those remaining filtered rows, the 2nd facet further filters, then the 3rd facet and so on.

If this is what you are seeing this is normal behavior.
If you can provide some sample data and steps to reproduce, than we can look into a potential bug around precedence ordering for numeric (or custom numeric) facets, but this is highly unlikely and instead you are simply seeing the above-described way facets work.

**Comment by thadguidry** [comment]
Hi @fabiolinus Can you describe for us how you would like the numeric facet to be enhanced?  Do you think controlling the binning size would help your use case as described in #5168 ?  What would the Numeric facet ideally look like and work in your above example?

**Comment by fabiolinus** [comment]
Sorry, I try to explain. Numeric facets should open with minimum and maximun. Sometimes I read values which are not identical to minimum and maximum; approximation? Then I can use something like in #5168, two numeric forms. I usually use a binning size equal to one. Another point is th relationship between numeric facet and one or more text facet. What is the distribution showed in the numeric facet? I see bins colored different according to subsets of data. I'd like to understand more about this point. Thanks a lot for your effort in OR,
Fabio


    Il lunedì 10 ottobre 2022 17:12:41 CEST, Thad Guidry ***@***.***> ha scritto:  
 
 


Hi @fabiolinus Can you describe for us how you would like the numeric facet to be enhanced? Do you think controlling the binning size would help your use case as described in #5168 ? What would the Numeric facet ideally look like and work in your above example?

**Comment by fabiolinus** [comment]
Thanks a lot. Let me thinking.
Fabio
    Il domenica 3 luglio 2022 14:41:55 CEST, Thad Guidry ***@***.***> ha scritto:  
 
 


There is no dependency. What you are witnessing is simply 2 facets that show values from filtered rows. Each facet can have a different filtering effect. However, there is a precedence order, in that if you move a facet into the top location of the facets panel, then that first facet will apply a filter to rows, then with those remaining filtered rows, the 2nd facet further filters, then the 3rd facet and so on.

If this is what you are seeing this is normal behavior.
If you can provide some sample data and steps to reproduce, than we can look into a potential bug around precedence ordering for numeric (or custom numeric) facets, but this is highly unlikely and instead you are simply seeing the above-described way facets work.

**Comment by ostephens** [comment]
> There is no dependency. What you are witnessing is simply 2 facets that show values from filtered rows. Each facet can have a different filtering effect.

I think this is the point that @fabiolinus is making (see also @wetneb comments).

- For a text facet the values listed are only taken from the remaining filtered rows
- For a numeric facet remains based on the full data set (although a subtle change of colour is used, the range remains the same)

There are at least a couple of issues here:
1. Is there a reason why we don't full recalculate & display the numeric facet based on the current filtered rows?
2. If the current behaviour is either unavoidable, or desirable some how, can we do a better job of differentiating between the included / non-included values? (doing this by colour alone is bad in terms of both general UX and accessibility, doing it by a very subtle change of colour is worse)

**Comment by thadguidry** [comment]
@ostephens James Home is the one who initially designed the histogram widget nearly from scratch with David's changes to the JSON and API.  I recall asking the question but don't remember the answer honestly, and I checked my email and notes and have nothing to offer as to why we don't fully recalculate based on current filtered rows.

It's definitely not desirable as the design was to have more added to the histogram later, such as setting scale with a MinMax algorithm (Stefano added Logarithmic transformation on Scatterplot but we never added it also to Range Facet along with others such as Square root, Cube root, etc.) And we never did add a Box plot to really help visualize better when data data is skewed.  As far as the grouping (binning) visualization, well we relied on the Text facet to accomplish alot since GREL had Floor which helped clump values if you added a new column to play with that.

We just never envisioned OpenRefine as a statistical tool and kinda still don't in my book.  But certainly improvements to the Range Facet and a new boxplot could help visualize outliers better, which was all the facets original intent to clean the skewed data.

I would definitely look towards other tools that have histograms such as PowerBI, Pandas in Jupyter Notebooks, and especially KNIME with Numeric Outliers is a personal favorite as inspiration for a better designed Range Facet and a new boxplot.

As to @fabiolinus concern about the visualization where the scale remains the same, this is true currently, but certainly doesn't have to stay that way, and for good frequency distribution visualization it certainly needs to improve, but then depending on what we do here, we begin to lean towards data science rather than data cleaning and analysis of outliers.  I'd choose visualizations that definitely help users filter better and leave it to other tools for the analysis if we could.

**Comment by fabiolinus** [comment]
Hi. Thanks a lot for your effort. I understand that OR is not a statistical tool or a visualization tool. A data scientist doesn't choose OR like first option, I think. OR is for data cleaning so it has to have minimal features to this goal. A boxplot is a good graphic to check for outliers. I'd like to use numeric facets to check if some values are missing in an interval, for example. I can have data in an complete column or a subset of this (for example for gender, group of ages, different levels and o on...). Thanks a lot again,
Fabio


    Il martedì 11 ottobre 2022 16:49:08 CEST, Thad Guidry ***@***.***> ha scritto:  
 
 


@ostephens James Home is the one who initially designed the histogram widget nearly from scratch with David's changes to the JSON and API. I recall asking the question but don't remember the answer honestly, and I checked my email and notes and have nothing to offer as to why we don't fully recalculate based on current filtered rows.

It's definitely not desirable as the design was to have more added to the histogram later, such as setting scale with a MinMax algorithm (Stefano added Logarithmic transformation on Scatterplot but we never added it also to Range Facet along with others such as Square root, Cube root, etc.) And we never did add a Box plot to really help visualize better when data data is skewed. As far as the grouping (binning) visualization, well we relied on the Text facet to accomplish alot since GREL had Floor which helped clump values if you added a new column to play with that.

We just never envisioned OpenRefine as a statistical tool and kinda still don't in my book. But certainly improvements to the Range Facet and a new boxplot could help visualize outliers better, which was all the facets original intent to clean the skewed data.

I would definitely look towards other tools that have histograms such as PowerBI, Pandas in Jupyter Notebooks, and especially KNIME with Numeric Outliers is a personal favorite as inspiration for a better designed Range Facet and a new boxplot.

As to @fabiolinus concern about the visualization where the scale remains the same, this is true currently, but certainly doesn't have to stay that way, and for good frequency distribution visualization it certainly needs to improve, but then depending on what we do here, we begin to lean towards data science rather than data cleaning and analysis of outliers. I'd choose visualizations that definitely help users filter better and leave it to other tools for the analysis if we could.
